registry/storage/revisionstore.go

```diff
@@ -131,16 +160,25 @@ func (rs *revisionStore) link(revision digest.Digest) error {
 	return rs.blobStore.link(revisionPath, revision)
 }
 
-// delete removes the specified manifest revision from storage.
+// delete creates a tombstone for the given digest revision.  The tombstone
+// is a link file which points to the deleted manifest.
 func (rs *revisionStore) delete(revision digest.Digest) error {
-	revisionPath, err := rs.pm.path(manifestRevisionPathSpec{
-		name:     rs.Name(),
-		revision: revision,
-	})
-
+	// don't allow deleting a blob that doesn't already exist
+	exists, err := rs.exists(revision)
 	if err != nil {
 		return err
 	}
 
-	return rs.driver.Delete(rs.repository.ctx, revisionPath)
+	if !exists {
+		return distribution.ErrUnknownManifestRevision{
+			Name:     rs.Name(),
+			Revision: revision,
+		}
+	}
+
+	if err := rs.tomb.putTombstone(rs.repository.ctx, rs.Name(), revision); err != nil {
+		return err
+	}
+
```

Comment:
Hello, @RichardScothern 
I tested this PR and found that when an image manifest is deleted, the related tags were not deleted.
Should the following lines be added here?
```diff
diff --git a/registry/storage/manifeststore.go b/registry/storage/manifeststore.go
index 841d45f..5d5589d 100644
--- a/registry/storage/manifeststore.go
+++ b/registry/storage/manifeststore.go
@@ -55,7 +55,19 @@ func (ms *manifestStore) Put(manifest *manifest.SignedManifest) error {
 // Delete removes the revision of the specified manfiest.
 func (ms *manifestStore) Delete(dgst digest.Digest) error {
 	ctxu.GetLogger(ms.repository.ctx).Debug("(*manifestStore).Delete")
-	return ms.revisionStore.delete(dgst)
+	if err := ms.revisionStore.delete(dgst); err != nil {
+		return err
+	}
+	tags, err := ms.Tags()
+	if err != nil {
+		return err
+	}
+	for _, tag := range tags {
+		if d, err := ms.tagStore.resolve(tag); err == nil && d == dgst {
+			err = ms.tagStore.delete(tag)
+		}
+	}
+	return err
 }
 
 func (ms *manifestStore) Tags() ([]string, error) {
```